Support for SCID Aliases

TheBlueMatt opened the pull request:
This is part 1/3 of the 0conf PR set, which only adds support for SCID aliaes using the new field in `funding_locked`. It does not yet support the relevant feature bit which is going to be a chunk more work as it will require fully implementing channel type negotiation, something we're currently missing (that will be PR 2). PR 3 will then add actual 0conf support, which isn't too bad after the first two land.

This and all the 0conf PRs will need a *very* careful review. I'd especially recommend reviewers scan through all our existing uses of `short_channel_id` and familiarize yourself with them either for this or later 0conf PRs as most uses of them will need to change.

Based on the 1.41.1 MSRV bump cause I'm lazy.

jkczyz commented:
> This and all the 0conf PRs will need a *very* careful review. I'd especially recommend reviewers scan through all our existing uses of `short_channel_id` and familiarize yourself with them either for this or later 0conf PRs as most uses of them will need to change.

Given this concern, could we make a `ShortChannelId` type that encapsulates the alias? Seems that would help prevent missed occurrences and possible future misuses. The interface would make it explicit which representation is used.

TheBlueMatt commented:
> Given this concern, could we make a ShortChannelId type that encapsulates the alias? Seems that would help prevent missed occurrences and possible future misuses. The interface would make it explicit which representation is used.

Hmm, we could make a newtype for it, I don't really have a strong feeling either way there, but I don't think we should make an enum, because there's not a ton (or any?) of places where we both (a) care which type it is, but (b) want to mix the two. In public APIs we mostly expose both and let the user figure it out, in internal stuff we mostly mix them and don't care which type it is.

TheBlueMatt commented:
Right, I think we need to be explicit about where we want it - external accessor(s) in `ChannelDetails` I think could make sense, dunno if we have to change the storage at all for that, though, just a util method could suffice. Internally I'm not sure there's a ton of cleanup we can do, but the public interface I think we could make simpler.

jkczyz commented:
The reason for using it internally is to avoid programmatic mistakes (both now and during further changes) and the need for reviewers to meticulously examine each use of `short_channel_id`. Maybe you could elaborate on what you mean by a util method as I'm not sure I understand.

TheBlueMatt commented:
Right, I think my point is that there isn't a good way to do it for internal stuff that isn't needless boilerplate that adds anything - there is, I believe, only exactly one place where both (a) we co-mingle the two types and (b) we care about knowing which type the thing is.

TheBlueMatt commented:
For *public* use, there's a few places now in eg lightning-invoice where we try to fetch the SCID alias and then fall back to SCID if there isn't one, that could be a util method on `ChannelDetails`, but its not exactly a super huge change.

jkczyz commented on lightning/src/ln/channelmanager.rs:
```diff
@@ -1380,6 +1393,23 @@ macro_rules! handle_error {
+			// If the channel was never confirmed on-chain prior to its closure, remove the
+			// outbound SCID alias we used for it from the collision-prevention set. While we
+			// generally want to avoid ever re-using an outbound SCID alias across all channels, we
+			// also don't want a counterparty to be able to trivially use up all our memory by
+			// simply opening a billion channels with us which are closed before we ever reach the
+			// funding stage.
```
Does this actually prevent that, though, if this map is inserted into on `open_channel`?

TheBlueMatt replied:
Right, I guess I should rephrase, I mean this prevents us *leaking*, but of course as long as the channels are pending we'll store channel info for them.

jkczyz commented on lightning/src/ln/channelmanager.rs:
```diff
@@ -1765,6 +1813,21 @@ impl<Signer: Sign, M: Deref, T: Deref, K: Deref, F: Deref, L: Deref> ChannelMana
+				outbound_scid_alias = fake_scid::Namespace::OutboundAlias.get_fake_scid(height, &self.genesis_hash, &self.fake_scid_rand_bytes, &self.keys_manager);
+			}
+			if self.outbound_scid_aliases.lock().unwrap().insert(outbound_scid_alias) { break; }
+			i += 1;
+			if i > 1_000_000 { panic!("Your RNG is busted or we ran out of possible outbound SCID aliases (which should never happen before we run out of memory to store channels"); }
```
Should this be done in `internal_open_channel`, too (in a DRY manner)?

TheBlueMatt replied:
Yea, this got added later and the thing keeps growing, I'll DRY it.

jkczyz commented:
> Right, I think my point is that there isn't a good way to do it for internal stuff that isn't needless boilerplate that adds anything - there is, I believe, only exactly one place where both (a) we co-mingle the two types and (b) we care about knowing which type the thing is.

Yeah, I think you're right. Took a more thorough look through the PR and spec change. This just seemed rather ominous on first read without fully grasping the spec change:

> This and all the 0conf PRs will need a _very_ careful review. I'd especially recommend reviewers scan through all our existing uses of `short_channel_id` and familiarize yourself with them either for this or later 0conf PRs as most uses of them will need to change.

jkczyz reviewed (commented):
Took an overall pass after better understanding the spec change. Wish there were a way to not publish nits so I can comment on them when I see them without publishing them. 😛

TheBlueMatt commented on lightning/src/ln/channelmanager.rs:
```diff
@@ -1187,6 +1193,13 @@ pub struct ChannelDetails {
 	/// The position of the funding transaction in the chain. None if the funding transaction has
 	/// not yet been confirmed and the channel fully opened.
 	pub short_channel_id: Option<u64>,
+	/// An optional [`short_channel_id`] alias for this channel, randomly generated by our
+	/// counterparty and usable in place of [`short_channel_id`] in invoice route hints. Our
+	/// counterparty will recognize the alias provided here in place of the [`short_channel_id`]
+	/// when they see a payment to be routed to us.
+	///
+	/// [`short_channel_id`]: Self::short_channel_id
+	pub inbound_scid_alias: Option<u64>,
```
Yea, I'll just comment and say you must use this instead of scid, if its set. no need to differentiate based on channel type, really.

TheBlueMatt commented on lightning/src/ln/channel.rs:
```diff
@@ -2141,24 +2161,44 @@ impl<Signer: Sign> Channel<Signer> {
+					// If they have sent updated points, funding_locked is always supposed to match
+					// their "first" point, which we re-derive here.
+					self.commitment_secrets.get_secret(INITIAL_COMMITMENT_NUMBER - 1)
+						.map(|secret| SecretKey::from_slice(&secret).ok()).flatten()
```
Actaully, we can unwrap every step here, we should have already validated the whole thing by the time we get here.

TheBlueMatt commented on lightning/src/ln/channel.rs:
```diff
@@ -4185,6 +4228,22 @@ impl<Signer: Sign> Channel<Signer> {
 		self.short_channel_id
 	}
 
+	/// Allowed in any state (including after shutdown)
+	pub fn get_latest_inbound_scid_alias(&self) -> Option<u64> {
+		self.latest_inbound_scid_alias
+	}
+
+	/// Allowed in any state (including after shutdown)
+	pub fn get_outbound_scid_alias(&self) -> u64 {
+		self.outbound_scid_alias
+	}
```
I mean I guess, what's the rationale, though? I figured this is one of those things that everyone is gonna have an opinion on and neither is really more or less readable, but maybe you have some specific criteria in mind that makes things clearer?

jkczyz replied:
Mostly to align with idiomatic rust. Maybe not so important for internal structs.

TheBlueMatt commented:
Rebased.

valentinewallace commented on lightning/src/ln/channelmanager.rs:
```diff
@@ -6774,6 +6836,32 @@ impl<'a, Signer: Sign, M: Deref, T: Deref, K: Deref, F: Deref, L: Deref>
+				return Err(DecodeError::InvalidValue);
+			}
+			if chan.is_usable() {
+				if short_to_id.insert(chan.outbound_scid_alias(), *chan_id).is_some() {
```
Do we intend to have two entries for channels with an `outbound_alias` in `short_to_id`? Could we add some docs for `short_to_id` and mention aliases?

valentinewallace commented on lightning/src/ln/priv_short_conf_tests.rs:
```diff
@@ -0,0 +1,286 @@
+}
+
+#[test]
+fn test_routed_scid_alias() {
```
Could we move the first payment's part of this test into `lightning-invoice/utils` to get some coverage on the invoice util changes too?

TheBlueMatt replied:
Hmmm, the change there is trivial, how about I just add an assertion to one of the tests in `utils` there to ensure the invoice is generated correctly?

valentinewallace commented on lightning/src/ln/channelmanager.rs:
```diff
@@ -4216,13 +4266,17 @@ impl<Signer: Sign, M: Deref, T: Deref, K: Deref, F: Deref, L: Deref> ChannelMana
 		let mut channel = Channel::new_from_req(&self.fee_estimator, &self.keys_manager, counterparty_node_id.clone(),
-				&their_features, msg, 0, &self.default_configuration, self.best_block.read().unwrap().height(), &self.logger)
+				&their_features, msg, 0, &self.default_configuration, self.best_block.read().unwrap().height(), &self.logger, outbound_scid_alias)
 			.map_err(|e| MsgHandleErrInternal::from_chan_no_close(e, msg.temporary_channel_id))?;
```
It looks like if this call fails we won't remove the new `outbound_scid_alias`?

valentinewallace commented on lightning/src/ln/channelmanager.rs:
```diff
@@ -4739,7 +4793,7 @@ impl<Signer: Sign, M: Deref, T: Deref, K: Deref, F: Deref, L: Deref> ChannelMana
 							raa_updates.finalized_claimed_htlcs,
 							chan.get().get_short_channel_id()
-								.expect("RAA should only work on a short-id-available channel"),
+								.unwrap_or(chan.get().outbound_scid_alias()),
```
If we're going to  make this change in this commit, should we also update the `get_short_channel_id()` unwrap in `handle_chan_restoration_locked`?

TheBlueMatt replied:
Oh, yea, lets just move this hunk to a later PR.

TheBlueMatt commented on lightning/src/ln/channelmanager.rs:
```diff
@@ -3175,12 +3254,9 @@ impl<Signer: Sign, M: Deref, T: Deref, K: Deref, F: Deref, L: Deref> ChannelMana
-											if let Some(short_id) = channel.get_short_channel_id() {
-												channel_state.short_to_id.remove(&short_id);
-											}
+											let mut channel = remove_channel!(self, channel_state, chan);
```
Grrr, let me redo all those callsites in a dependent PR.

TheBlueMatt replied:
I'm not sure that this is an issue in this PR? Its an issue once we go to 0conf, but we don't change the semantics of SCIDs here.

valentinewallace replied:
That's fair. On a semi-related note, I'm not sure we're compliant with this part of the spec:

```
 - if it returns a `channel_update`:
    - MUST set `short_channel_id` to the `short_channel_id` used by the incoming onion.
 ```

seems we'll always use `chan.get_short_channel_id()` in channel updates on onion error

TheBlueMatt replied:
Good catch. This is really nasty to resolve, so I may end up suggesting we leave this for the next PR as well, but I'll work on it and open up that PR today or so.

TheBlueMatt replied:
See #1351.

TheBlueMatt commented on lightning/src/ln/channelmanager.rs:
```diff
@@ -3454,7 +3530,7 @@ impl<Signer: Sign, M: Deref, T: Deref, K: Deref, F: Deref, L: Deref> ChannelMana
 			Ok(Some((update_fee, commitment_signed, monitor_update))) => {
 				if let Err(e) = self.chain_monitor.update_channel(chan.get_funding_txo().unwrap(), monitor_update) {
-					let (res, drop) = handle_monitor_err!(self, e, short_to_id, chan, RAACommitmentOrder::CommitmentFirst, false, true, Vec::new(), Vec::new(), Vec::new(), chan_id);
+					let (res, drop) = handle_monitor_err!(self, e, short_to_id, chan, RAACommitmentOrder::CommitmentFirst, chan_id, commitment_update_only);
```
Hmm, can you give a specific line number in reference to the top commit of the PR? line ~3730 is just `short_channel_id: None`?

valentinewallace replied:
Oops, I meant to refer to the other two places we generate `PaymentPathFailed`. In either case, is it possible that our own channel failed and the alias will be used as the `scid`? If so, those docs could use updating as well

TheBlueMatt replied:
Hmmm, dunno how that'd happen, we get the SCID from `onion_utils::process_onion_failure` which pulls from the route. I suppose its possible someone used an alias that is also one of our channels, if that's what you're asking. Or are you just saying it'd be bad if it were an alias and we'd punish the wrong channel if someone picks an alias that is also a real channel (but not the one we used)? I beileve we check that the channel's source pubkey matches the one from the route, so hopefully that is also impossible.

TheBlueMatt commented on lightning/src/ln/channel.rs:
```diff
@@ -695,6 +695,13 @@ pub(super) struct Channel<Signer: Sign> {
+	// on-chain transaction is ours.
+	// We only bother storing the most recent SCID alias at any time, though our counterparty has
+	// to store all of them.
+	latest_inbound_scid_alias: Option<u64>,
```
Hmm, that isn't the only place we use them post-0conf - they're general aliases, and someone could totally use them to route over a channel prior to confirmation.

TheBlueMatt replied:
I thought about this a bit more and I still kinda like `inbound` as the terminology here - it matches all the ways I'd think about it (its a value we received from our counterparty, and its also the value used for payments routed inbound towards us). I still kinda want to avoid terming it "route hint" cause users may use it outside of explicit invoice contexts (eg by handing someone a last-hop explicitly via some other protocol) and "route hint" sounds very invoice-y, even though the alternative uses I can see immediately are all similar concepts.

I guess I'm not quite sure why the inbound/outbound split is confusing - it seems to line up with all the things I can think of.

valentinewallace replied:
Sg, I guess not that it's confusing per se but I thought the rename would make it more obvious that the correct alias is being used in a given context. Fine to let it rest

TheBlueMatt commented on lightning/src/ln/channel.rs:
```diff
@@ -2151,6 +2162,15 @@ impl<Signer: Sign> Channel<Signer> {
+				// i.e. can be used for inbound payments and provided in invoices, but is not used
+				// when routing outbound payments.
+				self.latest_inbound_scid_alias = Some(scid_alias);
+			}
```
Hmm, probably not worth closing the channel for?

TheBlueMatt commented on lightning/src/ln/channel.rs:
```diff
@@ -6202,6 +6226,8 @@ impl<'a, Signer: Sign, K: Deref> ReadableArgs<(&'a K, u32)> for Channel<Signer>
 			workaround_lnd_bug_4006: None,
 
 			latest_inbound_scid_alias,
+			// Later in the ChannelManager deserialization phase we scan for channels and assign scid aliases if its missing
```
Hmmm, we don't currently do anything related to checking serialization from old versions, its definitely something we should do, but seems like a substantial project. cc #1089.

TheBlueMatt commented on lightning/src/ln/functional_tests.rs:
```diff
@@ -5916,7 +5916,7 @@ fn bolt2_open_channel_sending_node_checks_part1() { //This test needs to be on i
-	//Force duplicate channel ids
+	// Force duplicate randomness for every get-random call
 	for node in nodes.iter() {
 		*node.keys_manager.override_channel_id_priv.lock().unwrap() = Some([0; 32]);
```
Lets discuss in #1317.

ariard commented on lightning/src/ln/channel.rs:
```diff
@@ -695,6 +695,19 @@ pub(super) struct Channel<Signer: Sign> {
+	// outbound payments. These can be used in invoice route hints to provide privacy of which
+	// on-chain transaction is ours.
+	// We only bother storing the most recent SCID alias at any time, though our counterparty has
+	// to store all of them.
```
If the SCID alias goal is to mask the on-chain funding transaction, what's the interest to _rotate_ them ? I think one alias should be enough to achieve the purpose.

ariard commented on lightning/src/ln/channel.rs:
```diff
@@ -4195,6 +4239,22 @@ impl<Signer: Sign> Channel<Signer> {
+	}
+	/// Only allowed immediately after deserialization if get_outbound_scid_alias returns 0,
+	/// indicating we were written by LDK prior to 0.0.106 which did not set outbound SCID aliases.
+	pub fn set_outbound_scid_alias(&mut self, outbound_scid_alias: u64) {
```
I think you can precise the requirements of the given `outbound_scid_alias` if this channel utility is re-used in the future.

TheBlueMatt replied:
I have no idea what you mean here?

ariard replied:
What's the requirements about the `u64` ? Like at least to not give the _real_ channel SCID as an argument.

TheBlueMatt replied:
Recall `channel.rs` is internal-only, so this isn't for external consumption, and is pretty clear given it says `outbound_scid_alias` in the title :)

ariard commented on lightning/src/ln/channelmanager.rs:
```diff
@@ -660,6 +660,9 @@ pub(super) enum RAACommitmentOrder {
+	/// SCIDs (and outbound SCID aliases) to the real channel id. Outbound SCID aliases are added
+	/// here once the channel is available for normal use, with SCIDs being added once the funding
+	/// transaction is confirmed at the channel's required confirmation depth.
 	pub(super) short_to_id: HashMap<u64, [u8; 32]>,
```
If there is an existing outbound SCID alias, as the routing hop, should we forbid HTLC forward with the _real_ SCID ?

Otherwise a privacy attacker might probe the private channel existence by sending malicious `tlv_payload` built from collecting on-chain p2wsh..

TheBlueMatt replied:
That's coming in the next PR - support for the SCIDAlias channel type. We can't do it universally because the channel may be 0conf and public, in which case we want to relay over the real SCID and just use the alias for until we get 6 confs.

ariard commented on lightning/src/ln/channel.rs:
```diff
@@ -2151,24 +2172,44 @@ impl<Signer: Sign> Channel<Signer> {
 		}
 
+		if let Some(scid_alias) = msg.short_channel_id_alias {
+			if Some(scid_alias) != self.short_channel_id {
```
Should we sanitize further the scid alias ? Like ensuring it's not equal to the scid of other channels.

Also if we receive our counterparty's `funding_locked` before we have seen confirmation, the `short_channel_id` can be the temporary one. Then be replaced to the real one in `transactions_confirmed` ? It might be broken due to events unreliable ordering..

TheBlueMatt replied:
> Like ensuring it's not equal to the scid of other channels.

That doesn't matter, we never see the SCID alias anywhere, we just shove it in the invoice and its up to our counterparty to figure it out.

> Also if we receive our counterparty's funding_locked before we have seen confirmation, the short_channel_id can be the temporary one

Not sure what you mean here - in that case `self.short_channel_id` will be `None` so the comparison will always be true.

ariard replied:
> Not sure what you mean here - in that case self.short_channel_id will be None so the comparison will always be true.

Eh right, though if we receive a `funding_locked` before we have seen the confirmation, we don't set `latest_inbound_scid_alias` and our peer might never resend a `funding_locked` letting this field as `None` ?

TheBlueMatt replied:
No? If `self.short_channel_id` is `None` we'll always set the inbound scid alias field.

ariard reviewed (commented):
First parse, especially see comment about `funding_locked` unreliable ordering. 

The spec reference PR : https://github.com/lightning/bolts/pull/910

TheBlueMatt commented on lightning/src/ln/channel.rs:
```diff
@@ -695,6 +695,19 @@ pub(super) struct Channel<Signer: Sign> {
+	// Our counerparty can offer us SCID aliases which they will map to this channel when routing
+	// outbound payments. These can be used in invoice route hints to provide privacy of which
+	// on-chain transaction is ours.
+	// We only bother storing the most recent SCID alias at any time, though our counterparty has
```
> it leaks there was non-gossip communication between the sender and next hop

I'm confused what you mean by this - of course there's always non-gossip communication between the sender and the next-hop - they have a channel!

Further, whenever there is a public channel between any hop(s) the SCID alias doesn't help much - you've already announced the channel publicly on the gossip layer.

ariard replied:
> I'm confused what you mean by this - of course there's always non-gossip communication between the sender and the next-hop - they have a channel!

I meant, let's say you have A --> B --> C, where link BC is a SCID aliased, private channel.
C sends the SCID alias as route hint to A.
B receives the HTLC from A with a SCID alias as part of the `hop_payload`. B can infer that C is the final recipient of the HTLC.
If there was not a SCID alias used, B might not be able to infer with the same level of certainty that C is the HTLC payee. There could be another private link C -> after.

As said, I'm not sure if matters that much as multiple-hops routing through private channels might be rare on the network.

> Further, whenever there is a public channel between any hop(s) the SCID alias doesn't help much - you've already announced the channel publicly on the gossip layer.

You can hide to the HTLC sender that you're using _this_ public channel ? Like if you want to probe/jam a specific channel..

TheBlueMatt replied:
Ah, I see your point. Yea, I suppose that is clearer, but, indeed, as you note, for private channels it doesn't matter all that much. For public channels, after #1325, we'll not include any hints at all if we have some public channels so it shouldnt matter.

TheBlueMatt commented on lightning/src/ln/channelmanager.rs:
```diff
@@ -1383,6 +1414,24 @@ macro_rules! handle_error {
+			// also don't want a counterparty to be able to trivially cause a memory leak by simply
+			// opening a million channels with us which are closed before we ever reach the funding
+			// stage.
+			let alias_removed = $self.outbound_scid_aliases.lock().unwrap().remove(&$channel.outbound_scid_alias());
+			debug_assert!(alias_removed);
```
Not sure what you mean here? We're only removing if `get_short_channel_id` is `None`.

valentinewallace replied:
I think we should be removing even if it *isn't* `None`, at least pre-0conf?

TheBlueMatt replied:
I don't think so? See the comment two lines up - the only reason for this set is to try to ensure we don't re-use SCID aliases. Just because a channel was closed doesn't mean we want to re-use them later, at least until *much* later.

TheBlueMatt commented on lightning/src/ln/channelmanager.rs:
```diff
@@ -660,6 +660,9 @@ pub(super) enum RAACommitmentOrder {
+	/// SCIDs (and outbound SCID aliases) to the real channel id. Outbound SCID aliases are added
+	/// here once the channel is available for normal use, with SCIDs being added once the funding
+	/// transaction is confirmed at the channel's required confirmation depth.
 	pub(super) short_to_id: HashMap<u64, [u8; 32]>,
```
Hmm, not sure if we'd want to use `forward_htlcs` to address #1157, but I updated the docs.

TheBlueMatt commented on lightning-invoice/src/utils.rs:
```diff
@@ -162,7 +162,7 @@ where
 	let mut route_hints = vec![];
 	for channel in our_channels {
-		let short_channel_id = match channel.short_channel_id {
+		let short_channel_id = match channel.get_inbound_payment_scid() {
```
I'll leave that for the 0conf PR - we don't need to care right now cause first-hop SCID isn't communicated anywhere, its just an internal tracker for us, so its only needed after 0conf.

TheBlueMatt commented:
Addressed feedback, had to push a new cleanup commit at the start to deliver `channel_update`s to `ChannelManager`s in test, which allowed the invoice tests to include route hints.

valentinewallace commented:
@TheBlueMatt one comment here

jkczyz commented on lightning/src/ln/channelmanager.rs:
```diff
@@ -1797,6 +1797,23 @@ impl<Signer: Sign, M: Deref, T: Deref, K: Deref, F: Deref, L: Deref> ChannelMana
 		&self.default_configuration
 	}
 
+	fn create_insert_outbound_scid_alias(&self) -> u64 {
```
nit: create_or_insert_outbound_scid_alias

TheBlueMatt replied:
But it does both, so its not `_or_`.

jkczyz replied:
Ah, right lol. Maybe `_and_` or drop `insert_` if it is relevant to the caller? No strong opinion.

jkczyz commented on lightning/src/ln/channelmanager.rs:
```diff
@@ -1828,30 +1845,16 @@ impl<Signer: Sign, M: Deref, T: Deref, K: Deref, F: Deref, L: Deref> ChannelMana
 					let config = if override_config.is_some() { override_config.as_ref().unwrap() } else { &self.default_configuration };
 					Channel::new_outbound(&self.fee_estimator, &self.keys_manager, their_network_key, their_features,
-						channel_value_satoshis, push_msat, user_channel_id, config, height, outbound_scid_alias)?
+						channel_value_satoshis, push_msat, user_channel_id, config, self.best_block.read().unwrap().height(), outbound_scid_alias)?
```
I don't think I see any lock ordering concerns between `per_peer_state` and `best_block`, but perhaps docs could mention ordering.

TheBlueMatt replied:
Even better, let me just go update the debug lock stuff to support RwLock, will follow up with a separate PR.

TheBlueMatt replied:
#1352

jkczyz reviewed (commented):
Took another pass though. Remaining comments are fairly minor. Happy to see this go through and take a look at the next step.

TheBlueMatt commented on lightning/src/ln/functional_test_utils.rs:
```diff
@@ -693,6 +693,61 @@ pub fn create_announced_chan_between_nodes_with_value<'a, 'b, 'c, 'd>(nodes: &'a
 	(chan_announcement.1, chan_announcement.2, chan_announcement.3, chan_announcement.4)
 }
 
+pub fn create_unannounced_chan_between_nodes_with_value<'a, 'b, 'c, 'd>(nodes: &'a Vec<Node<'b, 'c, 'd>>, a: usize, b: usize, channel_value: u64, push_msat: u64, a_flags: InitFeatures, b_flags: InitFeatures) -> (msgs::FundingLocked, Transaction) {
+	let mut no_announce_cfg = test_default_channel_config();
+	no_announce_cfg.channel_options.announced_channel = false;
```
Hmm, all the existing utilities end up either calling `create_channel` itself or expecting to exchange `AnnouncementSignatures` at the end. I could rejigger the utilities above to separate those bits out, if you prefer? I don't really have a strong feeling between that vs having some code duplication in utilities right around each other.

jkczyz replied:
Don't feel too strongly. I think I already asked another contributor to do the same in another PR, heh.

TheBlueMatt replied:
Hmm, on second thought I'm kinda struggling to see how to do this - the top 7 lines could be split out into three lines and four lines, but that doesn't seem so useful. Then there's the funding generation, which almost can be combined, except that we rely on exchanging the funding_locked messages in order so that we only get one `channel_update` out at a time - we could swap for the `create_chan_between_nodes_with_value_confirm_first` helper, but then we're stick peeling apart multiple pending messages which takes more LoC than we save :(.

jkczyz replied:
No problem with passing on this. My concern was more with code repetition resulting in needing to modify two places rather than reducing lines of code.

TheBlueMatt replied:
Yea, I agree, its unclear how to resolve it in this case, though, without digging through a number of other tests and changing their util method calls, which is yet more diff in this PR :/

TheBlueMatt commented on lightning/src/ln/channelmanager.rs:
```diff
@@ -1281,6 +1287,15 @@ pub struct ChannelDetails {
+	/// Gets the SCID which should be used to identify this channel for inbound payments. This
+	/// should be used for providing invoice hints or in any other context where our counterparty
+	/// will forward a payment to us.
+	pub fn get_inbound_payment_scid(&self) -> Option<u64> {
```
Personally I'm kinda a fan of the `get_` prefix, they communicate the functions behavior (really, that it doesn't have side-effects) in a useful way. Its a bit less rust-idiomatic, but its a bit more java-idiomatic, and I dunno about swift/JS, so I'm not sure that's a strong reason.

jkczyz replied:
I guess my counter would be naming something that doesn't start with a verb indicates that it is a property.

TheBlueMatt replied:
In some idioms yes, but in Java isn't `get_` expected, even for a property-fetcher?

jkczyz replied:
Yeah, agreed. That is idiomatic for Java as I understand. Suppose we could always tag these somehow for the bindings generator, but no need to worry about this now.

valentinewallace reviewed (approved):
This looks good to me, done several passes now. Happy to start looking onto to part 2!

TheBlueMatt commented:
Squashed the fixup commits down.

TheBlueMatt commented:
Fuzz CI failure is an upstream regression, will fix separately. Kicked CI to hopefully make it pass here.

ariard commented:
Good reviewing again quickly rn and starting to look on PR 2.

ariard reviewed (commented):
ACK e4486fe
